src/data/error.rs

```diff
@@ -1,14 +1,59 @@
+use std::collections::VecDeque;
 use std::fmt;
-
 use thiserror::Error;
 
-use super::{Expr, Locatable, Location};
+use super::{Locatable, Location};
 
-pub type RecoverableResult<T = Expr, E = CompileError> = Result<T, (E, T)>;
+/// RecoverableResult is a type that represents a Result that can be recovered from.
+///
+/// See the [`Recover`] trait for more information.
+///
+/// [`Recover`]: trait.Recover.html
+pub type RecoverableResult<T, E = CompileError> = Result<T, (E, T)>;
 pub type CompileResult<T> = Result<T, CompileError>;
 pub type SemanticError = Locatable<String>;
 pub type CompileError = Locatable<Error>;
 
+/// ErrorHandler is a struct that hold errors generated by the compiler
+///
+/// An error handler is used because multiple errors may be generated by each
+/// part of the compiler, this cannot be represented well with Rust's normal
+/// `Result`.
+#[derive(Debug, Default, PartialEq)]
+pub(crate) struct ErrorHandler {
+    errors: VecDeque<CompileError>,
+}
+
+impl ErrorHandler {
+    /// Construct a new error handler.
+    pub(crate) fn new() -> ErrorHandler {
+        Default::default()
+    }
+
+    /// Add an error to the error handler.
+    pub(crate) fn push_err(&mut self, error: CompileError) {
+        self.errors.push_back(error);
+    }
+
+    pub(crate) fn pop_err(&mut self) -> Option<CompileError> {
+        self.errors.pop_front()
+    }
+
+    /// Checks if there are errors in the handler
+    pub(crate) fn is_empty(&self) -> bool {
+        self.errors.is_empty()
+    }
+}
+
+impl IntoIterator for ErrorHandler {
```

Comment:
Is there a reason this has to be `IntoIterator` instead of `Iterator`? You could implement `next()` with `self.errors.pop_front()`.

I guess that would allow calling `push_err()` after you've already looped over it ... I don't know that that's a bad thing though.

Reply:
This is only used for conveniently collecting into a Vec for some of the tests. So it doesn't matter to much.

Reply:
Changed it to Iterator